### PR TITLE
Remove pre from phase shifter and tap changer logics

### PR DESCRIPTION
**Pull request by adriguir**
The pre operators are removed when it makes sense for the logic on the phase shifter and tap changer.

In addition, some comments and typos in the associated documentation are corrected.

### CODE REVIEW

**Comment by casella** [comment]
@adriguir, @RomainLosseauRTE, based on our understanding, the use of `pre` is essential here. When the when clause is triggered, it changes some equations that affect the large algebraic system determining the phasors of the entire grid, which could in turn change the value of conditions such as `valueAboveMax`.

In this case, if you don't put `pre()`, you mean that the value of  `valueAboveMax` _after_ the event has been processed should be considered. This unnecessarily creates an algebraic loop involving mixed continuous and discrete equation that may be difficult or impossible to solve. 

If you put `pre()`, instead, you refer to the value immediately before the event was processed, which is a known value, so no nasty mixed algebraic loop. In fact, this event iteration could be considered as a way to solve the nasty mixed nonlinear equations with a fixed point scheme, which is however clearly defined in the code, and not left to the backend to find out.

Note that if the switching causes conditions such as `valueAboveMax` to change, then event iteration is triggered (see [Section 3.7.3.1](https://specification.modelica.org/v3.4/Ch3.html#pre) and [Appendix C](https://specification.modelica.org/v3.4/A3.html) of the Modelica Specification).

Summing up, using pre() here is not a dirty trick like using the memory block in Simulink, but is rather the correct declarative way to formulate the problem in Modelica, i.e., referring to the values of conditions immediatly before the event is processed.

So, I strongly suggest _not_ to merge this PR, unless there are some other reasons that I am missing to do so.

@AndreaBartolini, can you comment on this?

**Comment by casella** [comment]
For the future, I would also suggest not to mix up different things in the same PR (e.g. removing pre() and fixing comments), so that each can be considered and processed separately

**Comment by AndreaBartolini** [comment]
The idea is exactly as Francesco has explained, in my experience the pre() operator is necessary for all the signals (i.e. variables) which are trigger condition and, at the same time, can change as consequence of the triggered condition itself, creating de facto an algebraic loop.
This is prevented for the "state" variable, because the elsewhen clause force the event iteration to go to the next step before to re-trigger another condition on the same "state" variable.
Therefore, the pre() operator is redundant just for the "state" variable, but it seems that it has been left unchanged in this case...

**Comment by adriguir** [comment]
Regarding the pre() use or not, I agree that it leads to solve a large algebraic problem after the event to reconsider the other values of "valueAboveMax" which could be quite a costy treatment and a hard problem to solve.
Nevertheless I get the impression that I can get in trouble if I have the following sequence that could be realistic: at t = 10, I get a line tripping leading to a valueAboveMax for tap1 at t+h and a valueAboveMax for tap2 at t + h' (h' > h but < to the time step). 30 s later (t + 30 + h), tap1 increases its tap which leads to releasing the constraint on valueAboveMax for tap2. When I get to t + 30 + h', pre(valueAboveMax) on tap2 will still be true while it shouldn't because the constraint was released by the previous change on tap1 so tap2 shouldn't move. I perhaps miss something on the event handling in Modelica tools (I have to admit that I never took time to deeply understand this part).

I can try to set up such a pathological case, if it works with the pre() then it's fine. Otherwise, we will see.

In the meanwhile, we can close the PR.

**Comment by adriguir** [comment]
@AndreaBartolini if I get it right, yes indeed, the condition in the when could only change once during one event iteration as the pre() operator is still there for the "state" (variable name) variable to keep the idea of transition.

My point is to immediately propagate the effect of tap changer 1 move on tap changer 2 for example.

**Comment by casella** [comment]
> Nevertheless I get the impression that I can get in trouble if I have the following sequence that could be realistic: at t = 10, I get a line tripping leading to a valueAboveMax for tap1 at t+h and a valueAboveMax for tap2 at t + h' (h' > h but < to the time step). 30 s later (t + 30 + h), tap1 increases its tap which leads to releasing the constraint on valueAboveMax for tap2. When I get to t + 30 + h', pre(valueAboveMax) on tap2 will still be true

Why should it? 

According to the [specification](https://specification.modelica.org/v3.4/Ch3.html#event-related-operators-with-function-syntax), 
- `pre(y)` returns the “left limit” y(t_pre) of variable y(t) at a time instant t. At an event instant, y(t_pre) is the value of y after the last event iteration at time instant t

so `pre(y)` it is not the previous value in a sequence of discrete events, as you seem to implicitly assume in your line of reasoning, but just the value immediately before the event was triggered or, in case of event iterations, the value a the end of the previous event iteration.

In this case, `valueAboveMax` is defined by this equation in class `TransformerWithPhaseShifterInterval`
```
valueAboveMax = portB.I > IMax;
```
At t = 30 + h, tap1 is changed, so that condition immediately becomes false. When you get to t = 30 + h', `pre(valueAboveMax)` will still be false, because it is the value of that expression immediately before the event, hence a time t = 30 + h' - epsilon

@adriguir, do I miss something?

**Comment by adriguir** [comment]
No, that's fine. Indeed, it was my understanding of the pre() operator that was wrong.

**Comment by casella** [comment]
I took me several years to really fully understand how it works, don't worry 😄
